### PR TITLE
gcp - add metric alert filter

### PR DESCRIPTION
**Pull request by hiteshmck**
Current in gcp there is no way to know if there are any matching filters for a metric. Created a filter named "alerts" that can be used to achieve this.

### CODE REVIEW

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    schema = type_schema('alerts')
+    required_keys = {}
+    permissions = ("monitoring.alertPolicies.list",)
+
+    def process(self, resources, event=None):
+        self.findings_list = self.get_findings(resources)
+        matched = [r for r in resources if self.process_resource(r)]
+        return matched
+
+    def get_findings(self, resources):
```
[the remark this answers is not in the document]
Comment (hiteshmck) [reply tools/c7n_gcp/c7n_gcp/filters/alerts.py]: Removed. Thx

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+GCP Filter to find Alerts for a Log Metric Filter
+"""
+from c7n.filters.core import ValueFilter
+from c7n.utils import local_session, type_schema
+from c7n_gcp.provider import resources as gcp_resources
+import jmespath
+
+
+class AlertsFilter(ValueFilter):
+    """GCP Filter to find Alerts for a Log Metric Filter.
```
[the remark this answers is not in the document]
Comment (hiteshmck) [reply tools/c7n_gcp/c7n_gcp/filters/alerts.py:13]: Registration filter included now. Thx for the feedback.

**Comment by hiteshmck** [comment]
@thisisshi Thx for the feedback. I have incorporated them. Please review when you get a chance.

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        'metric.type=\"logging.googleapis.com/user/" + resource_name + "\"')]"
+        result = jmespath.search(search_string, self.findings_list)
+
+        if not any(result):
+            return False
+        else:
+            return True
+
+    @classmethod
+    def register_resources(klass, registry, resource_class):
```
Comment (thisisshi) [inline tools/c7n_gcp/c7n_gcp/filters/alerts.py:59]: this doesnt really make much sense here, just register the filter directly to the log-project-metric resource itself if it's only going to be used on a single resource type

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        findings_paged_list = list(client.execute_paged_query('list',
+            {'name': 'projects/' + project, **query_params}))
+        findings_list = []
+        for findings_page in findings_paged_list:
+            if findings_page.get('alertPolicies'):
+                findings_list.extend(findings_page['alertPolicies'])
+        return findings_list
+
+    def process_resource(self, resource):
+        resource_name = resource.get('name')
```
[the remark this answers is not in the document]
Comment (thisisshi) [reply tools/c7n_gcp/c7n_gcp/filters/alerts.py:48]: in that case, would be better to do a direct reference here

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        search_string = "[*].conditions[?contains(conditionThreshold.filter,\
+        'metric.type=\"logging.googleapis.com/user/" + resource_name + "\"')]"
+        result = jmespath.search(search_string, self.findings_list)
+
+        if not any(result):
```
Comment (thisisshi) [inline tools/c7n_gcp/c7n_gcp/filters/alerts.py:53]: i dont think is sufficient to say that alerts are on, we should also check that the state of the alert's "enabled" flag is true

**Review by thisisshi — CHANGES_REQUESTED** [review]

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (hiteshmck) [reply tools/c7n_gcp/c7n_gcp/filters/alerts.py]: Thx, I put a direct reference now.

File: tools/c7n_gcp/c7n_gcp/filters/alerts.py (hunk 3 above)
Comment (hiteshmck) [reply tools/c7n_gcp/c7n_gcp/filters/alerts.py]: Thx, used decoraters  to register filter only for single resource.

**Comment by hiteshmck** [comment]
@thisisshi I incorporated changes per your comments. Please review.

**Review by thisisshi — APPROVED** [review]
thanks, lgtm

**Pull request merged**